diff --git a/recompui/src/renderer/ui_renderer.cpp b/recompui/src/renderer/ui_renderer.cpp
--- a/recompui/src/renderer/ui_renderer.cpp
+++ b/recompui/src/renderer/ui_renderer.cpp
@@ -573,6 +573,8 @@ public:
             copy_command_list_->copyTextureRegion(
                 plume::RenderTextureCopyLocation::Subresource(texture.get()),
                 plume::RenderTextureCopyLocation::PlacedFootprint(copy_buffer_.get(), RmlTextureFormat, source_dimensions.x, source_dimensions.y, 1, row_width));
+            copy_command_list_->barriers(plume::RenderBarrierStage::GRAPHICS,
+                plume::RenderTextureBarrier(texture.get(), plume::RenderTextureLayout::SHADER_READ));
             
             // End the command list, execute it and wait.
             copy_command_list_->end();
